src/Reserve.sol

```diff
@@ -37,44 +122,146 @@ contract Reserve is IReserve, Ownable {
         _;
     }
 
-    function withdraw(
-        IERC20 token,
-        address to,
-        uint256 amt
-    ) public override onlyUser {
-        uint256 balance = withdrawable[token];
-        require(balance >= amt, "Reserve: withdrawal over balance");
-        withdrawable[token] = balance - amt;
-        require(token.transfer(to, amt), "Reserve: transfer failed");
-        emit Withdrawn(msg.sender, token, to, amt);
+    /// @notice Sets a plugin for a given token.
+    /// All future deposits and withdrawals of that token will be made using that plugin.
+    /// All currently deposited tokens of that type will be withdrawn from the plugin previously
+    /// set for that token and deposited into the new one.
+    /// If no plugin has been set, funds are deposited from the reserve itself.
+    /// If no plugin is being set, funds are deposited into the reserve itself.
+    /// Callable only by the current owner.
+    /// @param token The used token.
+    /// @param newPlugin The new plugin address. `NO_PLUGIN` if no plugin should be used.
+    function setPlugin(IERC20 token, IReservePlugin newPlugin) public onlyOwner {
+        IReservePlugin oldPlugin = plugins[token];
+        plugins[token] = newPlugin;
+        uint256 amt = deposited[token];
+        _beforeWithdrawal(token, oldPlugin, amt);
+        _transfer(token, _pluginAddr(oldPlugin), _pluginAddr(newPlugin), amt);
+        _afterDeposition(token, newPlugin, amt);
+        emit PluginSet(msg.sender, token, oldPlugin, newPlugin);
     }
 
+    /// @notice Deposits funds into the reserve.
```

Comment:
It would be nice to add the following method.

```solidity
function surplus(IERC20 erc20) external view returns (int surplus) {
 IReservePlugin plugin = plugins[token];
// could be negative as well
 returns plugin.balance() - deposited[token];
}

```

Currently, we don't have a strategy what should happen with a `surplus` but I think for now we should assume to governance should be able to take the `surplus` with a governance call forced `withdraw`

Reply:
However, it might be too much work to start a governance process each time.

Instead i could imagine the following. The `onlyOwner` can set a `surplusReceiver` for a specific token.

Together with a public callable `surplusHarvest` method, which pays out the difference between `plugin.balance() - deposited[token]` if `surplusReceiver` is set.

Reply:
Yes, we have very high uncertainty in the design phase. Therefore, I agree to keep it as generic as possible.

>  Who can claim the surplus?

This was my idea with the setting a `surplusReceiver`.  I can see you reasoning. I think to keep the `surplus` completely out of the system also works.

We only need to be aware of the limitations. That a `plugin` migration would not migrate the surplus. The plugin needs to take care of it.

I think most likely, we will end up with common surplus strategy for all plugins or ERC20.

For example the entire surplus should go to this Multi-Sig or maybe later you want to change it that a specific reward contract should receive it.

It might be beneficial to change them one in all step. Which we might loss here. 

However, I don't see it as that critical because most likely we won't have that many.

Reply:
> a `plugin` migration would not migrate the surplus

Yes, that's a drawback, it makes the plugin migration a more complex operation. We must decide what to do with the surplus and if we choose to move it to the new plugin, then we must organize it outside of the `Reserve`.

> It might be beneficial to change them one in all step.

Nothing is stopping us from registering a single smart contract for multiple ERC-20 tokens, we can then do all kinds of batch changes. We can also modularize further and have multiple plugins but all implementing a common surplus withdrawal API callable by a contract deciding what to do with these funds.

Reply:
Ok, so let's keep this PR's design.